source/common/common/mem_block_builder.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+
+// Manages a block of raw memory for objects of type T. T is generally expected
+// to be a POD, where it makes sense to memcpy over it. This class carries extra
+// member variables for tracking size, and a write-pointer to support safe
+// appends.
+//
+// MemBlockBuilder is used to safely write blocks of data into a memory
+// buffer. Due to two extra member variables, it is not optimal for storing in
+// data structures. The intended usage model is to release the raw assembled
+// memory block from the MemBlockBuilder for efficient storage.
+//
+// The goal for this class is to provide a usage model to replace direct usage
+// of memcpy with a pattern that is easy to validate for correctness by
+// inspection, asserts, and fuzzing, but when compiled for optimization is
+// roughly as efficient as raw memcpy.
+template <class T> class MemBlockBuilder {
+public:
+  // Constructs a MemBlockBuilder allowing for 'capacity' instances of T.
+  explicit MemBlockBuilder(uint64_t capacity)
+      : data_(std::make_unique<T[]>(capacity)), capacity_(capacity), capacity_remaining_(capacity),
+        write_ptr_(data_.get()) {}
+  MemBlockBuilder() : capacity_(0), capacity_remaining_(0), write_ptr_(nullptr) {}
+
+  /**
+   * Populates (or repopulates) the MemBlockBuilder to make it the specified
+   * capacity. This does not have resize semantics; when populate() is called any
+   * previous contents are erased.
+   *
+   * @param capacity The number of memory elements to allocate.
+   */
+  void populate(uint64_t capacity) {
```

Comment:
nit: I would probably just merge this with `reset` below and have reset take a capacity size. If capacity is 0, it can do what reset currently does? WDYT?

Reply:
PTAL -- I made a populateHelper method, to which you pass the capacity and the unique_ptr, to avoid a conditional, which disturbs me (even though I know it would be optimized out)...